File: docs/snippets/all/reference/dataframe_query.rs
//! Query and display the first 10 rows of a recording.

use rerun::{
    dataframe::{QueryEngine, QueryExpression, SparseFillStrategy, TimelineName},
    external::re_format_arrow::format_record_batch,
    ChunkStoreConfig,
};

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args = std::env::args().collect::<Vec<_>>();

    let path_to_rrd = &args[1];
    let timeline = TimelineName::log_time();

    let engines = QueryEngine::from_rrd_filepath(&ChunkStoreConfig::DEFAULT, path_to_rrd)?;

    let Some((_, engine)) = engines.first_key_value() else {
        return Ok(());
    };

    let query = QueryExpression {
        filtered_index: Some(timeline),
        sparse_fill_strategy: SparseFillStrategy::LatestAtGlobal,
        ..Default::default()
    };

    let query_handle = engine.query(query.clone());
    for row in query_handle.batch_iter().take(10) {
        // Each row is a `RecordBatch`, which can be easily passed around across different data ecosystems.
        println!("{}", format_record_batch(&row));
    }

    Ok(())
}

File: docs/snippets/all/reference/dataframe_query.py
"""Query and display the first 10 rows of a recording."""

import sys
from pathlib import Path

import rerun as rr

path_to_rrd = Path(sys.argv[1])

with rr.server.Server(datasets={"dataset": [path_to_rrd]}) as server:
    dataset = server.client().get_dataset("dataset")
    df = dataset.reader(index="log_time")

    table = df.to_arrow_table()
    top_ten = table.slice(0, min(10, table.num_rows))
    print(top_ten)
